### DISCUSSION TITLE
Cannot return HashMap value.

### DISCUSSION THREAD

**Discussion by irishgreencitrus** [mlua-rs/mlua#210] [Q&A] [answered]
I have two structs. One holds a HashMap of a String to the other. When I try to implement a method getting the value of the hashmap from a string, it only returns a borrow, and I cannot implement the `Clone` trait due to the type containing a RegistryKey. Any help is appreciated.
```rs
struct PackageManager {
    install_callback: RegistryKey,
    packages: Vec<String>,
}
struct PackageScope {
    builtins: HashMap<String, PackageManager>,
    user: String,
}
impl UserData for PackageManager { ... }
impl UserData for PackageScope {
     fn add_methods<'lua, M: LuaUserDataMethods<'lua, Self>>(methods: &mut M) {
          methods.add_meta_method(LuaMetaMethod::Index, |lua, this: &Self, key: String| {
               let ret = this.builtins.get(&key);
               return match ret {
                   Some(val) => Ok(val),
                   None      => Err(LuaError::RuntimeError("Invalid key".to_string()))
               };
           });
       }
}
```

**Comment by khvzak** [comment, marked as answer]
There are few options:
1) Store PackageManager inside `Rc<RefCell>`:
```rust
struct PackageScope {
    builtins: HashMap<String, Rc<RefCell<PackageManager>>>,
    user: String,
}
``` 
Then you can send it to Lua (cloned by reference).

2) Store `RegistryKey` as `Rc<RegistryKey>`, this would allow to implement `Clone` for `PackageManager`

3) Store `PackageManager` directly inside Lua and refer to each instance via `RegistryKey`, eg:
```rust
struct PackageScope {
    builtins: HashMap<String, RegistryKey>,
    user: String,
}
```

**Reply by irishgreencitrus** [reply]
> There are few options:
> 
> 1) Store PackageManager inside `Rc<RefCell>`:
> 
> ```rust
> 
> struct PackageScope {
> 
>     builtins: HashMap<String, Rc<RefCell<PackageManager>>>,
> 
>     user: String,
> 
> }
> 
> ``` 
> 
> Then you can send it to Lua (cloned by reference).
> 
> 
> 
> 2) Store `RegistryKey` as `Rc<RegistryKey>`, this would allow to implement `Clone` for `PackageManager`
> 
> 
> 
> 3) Store `PackageManager` directly inside Lua and refer to each instance via `RegistryKey`, eg:
> 
> ```rust
> 
> struct PackageScope {
> 
>     builtins: HashMap<String, RegistryKey>,
> 
>     user: String,
> 
> }
> 
> ```
Thanks so much for the help!
